File: leetcode/interview17.21_VolumeOfHistogram/lc17_21.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <= 2:
            return 0
        left = 0
        right = len(height) - 1
        high = 1
        volume = 0
        while left <= right: # double pointer
            while left <= right and height[left] < high:
                left += 1
            while left <= right and height[right] < high:
                right -= 1
            volume += right - left + 1 # volume between interval greater (or equal) than 'high' could contain water
            high += 1
        return volume - sum(height)

File: leetcode/interview17.21_VolumeOfHistogram/lc17_21.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int start = 0;
        int end;
        int answer = 0;
        int n = height.size();
        while (start < n - 2) {
            end = start + 1;
            if (height[end] >= height[start]) {
                start++;
            } else {
                while (end < n && height[end] < height[start]) end++;
                if (end < n) {
                    for (int i = start + 1; i < end; ++i) {
                        answer += height[start] - height[i];
                        height[i] = height[start]; // fill water
                    }
                    start = end;
                } else {
                    start++;
                }
            }
        }
        start = n - 1; // scan from right, case like [4, 2, 3] would be missed
        while (start > 1) {
            end = start - 1;
            if (height[end] >= height[start]) {
                start--;
            } else {
                while (end >= 0 && height[end] < height[start]) end--;
                if (end >= 0) {
                    if (height[end] > height[start]) {
                        for (int i = start - 1; i > end; --i) answer += height[start] - height[i];
                    }
                    start = end;
                } else {
                    start--;
                }
            }
        }
        return answer;
    }
};
